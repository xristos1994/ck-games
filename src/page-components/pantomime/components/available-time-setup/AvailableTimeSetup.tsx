import React, { FC, ReactElement } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { classnames, compose } from '@utils/component-utils';
import { withTranslation, ITranslate } from '@models/i18n/hoc';
import { Button } from '@components';
import {
  availableTime,
  availableTimes,
  canGoBack
} from '@models/pantomime/props';
import {
  setAvailableTime,
  availableTimeSetupSubmit,
  goBack
} from '@models/pantomime/actions';
import { IState } from '@models/interfaces';

import styles from './styles.module.css';

interface IProps {
  availableTime: number;
  availableTimes: number[];
  canGoBack: boolean;
  goBack: () => void;
  setAvailableTime: (availableTime: number) => void;
  availableTimeSetupSubmit: () => void;
  t: ITranslate;
}

const _AvailableTimeSetup: FC<IProps> = ({
  availableTime,
  availableTimes,
  canGoBack,
  setAvailableTime,
  availableTimeSetupSubmit,
  goBack,
  t
}): ReactElement => {
  const onAvailableTimeChange = (e: React.ChangeEvent<HTMLSelectElement>): void => {
    setAvailableTime(Number(e.target.value));
  };

  return (
    <div className={classnames(styles.availableTimeSetupContainer)}>
      <div className={classnames(styles.availableTimeSetupTitle)}>
        {t('Choose Available Time')}
      </div>
      <select
        className={classnames(styles.availableTimeTargetInput)}
        value={availableTime}
        onChange={() => onAvailableTimeChange}
      >
        {availableTimes.map(time => (
          <option key={time} value={time}>
            {`${time} ${t('Seconds')}`}
          </option>
        ))}
      </select>
      <Button
        onClick={() => availableTimeSetupSubmit()}
        className={classnames(styles.availableTimeTargetSetupSubmitButton)}
      >
        {t('CONTINUE')}
      </Button>
      {canGoBack && (
        <Button
          onClick={() => goBack()}
          className={classnames(styles.backButton)}
        >
          {t('BACK')}
        </Button>
      )}
    </div>
  );
};

const AvailableTimeSetup = compose(
  connect(
    createStructuredSelector<
      IState,
      {
        availableTime: IProps['availableTime'];
        availableTimes: IProps['availableTimes'];
        canGoBack: IProps['canGoBack'];
      },
      {
        availableTime: IProps['availableTime'];
        availableTimes: IProps['availableTimes'];
        canGoBack: IProps['canGoBack'];
      }
    >({
      availableTime,
      availableTimes,
      canGoBack
    }),
    { setAvailableTime, availableTimeSetupSubmit, goBack }
  ),
  withTranslation
)(_AvailableTimeSetup);

export { AvailableTimeSetup };
