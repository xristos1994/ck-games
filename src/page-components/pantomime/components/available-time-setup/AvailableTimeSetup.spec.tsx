import React, { MouseEventHandler, ReactChild } from 'react';
import renderer from 'react-test-renderer';
import { _AvailableTimeSetup as AvailableTimeSetup } from './AvailableTimeSetup';

const Button = (
  { children, onClick, className='' } :
  { children: ReactChild | ReactChild[]; onClick: MouseEventHandler<HTMLButtonElement>; className: string }
) => {
  return (
    <button onClick={onClick} className={className || undefined}>
      {children}
    </button>
  );
};

jest.mock('@components', () => ({
  __esModule: true,
  Button: Button
}));

const availableTimes = [30, 60, 90, 120];
const t = jest.fn().mockImplementation((label: string, param?: string[]) => label + (param ? param[0] : ''));
const setAvailableTime = jest.fn();
const availableTimeSetupSubmit = jest.fn();
const goBack = jest.fn();

describe('page-components/Pantomime/AvailableTimeSetup', () => {
  it('renders correctly', () => {
    let tree = renderer.create(
      <AvailableTimeSetup
        availableTime={availableTimes[0]}
        canGoBack={true}
        availableTimes={availableTimes}
        goBack={goBack}
        setAvailableTime={setAvailableTime}
        availableTimeSetupSubmit={availableTimeSetupSubmit}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <AvailableTimeSetup
        availableTime={availableTimes[1]}
        canGoBack={true}
        availableTimes={availableTimes}
        goBack={goBack}
        setAvailableTime={setAvailableTime}
        availableTimeSetupSubmit={availableTimeSetupSubmit}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has the right event handlers', () => {
    const tree = renderer.create(
      <AvailableTimeSetup
        availableTime={availableTimes[0]}
        canGoBack={true}
        availableTimes={availableTimes}
        goBack={goBack}
        setAvailableTime={setAvailableTime}
        availableTimeSetupSubmit={availableTimeSetupSubmit}
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        t={t}
      />
    );

    const selectOptions = tree.root.findByType('select');
    selectOptions.props.onChange({ target: { value: availableTimes[0] } });
    expect(setAvailableTime).toHaveBeenCalledTimes(1);
    expect(setAvailableTime).toHaveBeenLastCalledWith(availableTimes[0]);
  });
});
