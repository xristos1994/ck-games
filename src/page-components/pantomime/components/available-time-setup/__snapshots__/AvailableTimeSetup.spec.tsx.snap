// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page-components/Pantomime/AvailableTimeSetup renders correctly 1`] = `
<div
  className="availableTimeSetupContainer"
>
  <div
    className="availableTimeSetupTitle"
  >
    Choose Available Time
  </div>
  <select
    className="availableTimeTargetInput"
    onChange={[Function]}
    value={30}
  >
    <option
      value={30}
    >
      30 Seconds
    </option>
    <option
      value={60}
    >
      60 Seconds
    </option>
    <option
      value={90}
    >
      90 Seconds
    </option>
    <option
      value={120}
    >
      120 Seconds
    </option>
  </select>
  <button
    className="availableTimeTargetSetupSubmitButton"
    onClick={[MockFunction]}
  >
    CONTINUE
  </button>
  <button
    className="backButton"
    onClick={[MockFunction]}
  >
    BACK
  </button>
</div>
`;

exports[`page-components/Pantomime/AvailableTimeSetup renders correctly 2`] = `
<div
  className="availableTimeSetupContainer"
>
  <div
    className="availableTimeSetupTitle"
  >
    Choose Available Time
  </div>
  <select
    className="availableTimeTargetInput"
    onChange={[Function]}
    value={60}
  >
    <option
      value={30}
    >
      30 Seconds
    </option>
    <option
      value={60}
    >
      60 Seconds
    </option>
    <option
      value={90}
    >
      90 Seconds
    </option>
    <option
      value={120}
    >
      120 Seconds
    </option>
  </select>
  <button
    className="availableTimeTargetSetupSubmitButton"
    onClick={[MockFunction]}
  >
    CONTINUE
  </button>
  <button
    className="backButton"
    onClick={[MockFunction]}
  >
    BACK
  </button>
</div>
`;
